packages/framework/oldest-client-observer/src/oldestClientObserver.ts

```diff
@@ -0,0 +1,125 @@
+ * @remarks
+ * ### Creation
+ *
+ * The `OldestClientObserver` constructor takes an `IOldestClientObservable`.  This is most easily satisfied with
+ * either an `IContainerRuntime` or an `IFluidDataStoreRuntime`:
+ *
+ * ```typescript
+ * // E.g. from within a BaseContainerRuntimeFactory:
+ * protected async containerHasInitialized(runtime: IContainerRuntime) {
+ *     const oldestClientObserver = new OldestClientObserver(runtime);
+ *     // ...
+ * }
+ * ```
+ *
+ * ```typescript
+ * // From within a DataObject
+ * protected async hasInitialized() {
+ *     const oldestClientObserver = new OldestClientObserver(this.runtime);
+ *     // ...
+ * }
+ * ```
+ *
+ * ### Usage
+ *
+ * To check if the local client is the oldest, use the `isOldest()` method.
+ *
+ * ```typescript
+ * if (oldestClientObserver.isOldest()) {
+ *     console.log("I'm the oldest");
+ * } else {
+ *     console.log("Someone else is older");
+ * }
+ * ```
+ *
+ * ### Eventing
+ *
+ * `OldestClientObserver` is an `EventEmitter`, and will emit events when the local client becomes the oldest and when
+ * it is no longer the oldest.
+ *
+ * ```typescript
+ * oldestClientObserver.on("becameOldest", () => {
+ *     console.log("I'm the oldest now");
+ * });
+ *
+ * oldestClientObserver.on("lostOldest", () => {
+ *     console.log("I'm not the oldest anymore");
+ * });
+ * ```
+ */
+export class OldestClientObserver extends TypedEventEmitter<IOldestClientObserverEvents>
+    implements IOldestClientObserver {
+    private readonly quorum: IQuorumClients;
+    private currentIsOldest: boolean = false;
+    constructor(private readonly observable: IOldestClientObservable) {
+        super();
+        this.quorum = this.observable.getQuorum();
+        this.currentIsOldest = this.computeIsOldest();
+        this.quorum.on("addMember", this.updateOldest);
+        this.quorum.on("removeMember", this.updateOldest);
+        observable.on("connected", this.updateOldest);
+        observable.on("disconnected", this.updateOldest);
+    }
+
+    public isOldest(): boolean {
+        return this.currentIsOldest;
+    }
+
+    private readonly updateOldest = () => {
+        const oldest = this.computeIsOldest();
+        if (this.currentIsOldest !== oldest) {
+            this.currentIsOldest = oldest;
+            if (oldest) {
+                this.emit("becameOldest");
+            } else {
+                this.emit("lostOldest");
+            }
+        }
+    };
+
+    private computeIsOldest(): boolean {
+        // If the container is detached, we are the only ones that know about it and are the oldest by default.
+        if (this.observable.attachState === AttachState.Detached) {
+            return true;
+        }
+
+        // If we're not connected we can't be the oldest connected client.
+        if (!this.observable.connected) {
+            return false;
+        }
+
+        assert(this.observable.clientId !== undefined, 0x1da /* "Client id should be set if connected" */);
```

Comment:
@ChumpChief, is there a precedent for moving classes to a new package? My idea here was to stage the changes by deprecating the old class but looks like duplicate assert codes are an issue. Since it's only one assert, I could change one of the codes and leave a comment saying something like, "old code is 0x1da". What do you think?

Reply:
My concern is that I know that `OldestClientObserver` is being used by partner teams, so I thought it would be best to stage in this case

Reply:
Experimental packages are our space to explore new ideas and let our partners try them out - we don't offer back-compat promises on them (we even reserve the right to delete them entirely) and so our guidance has been to never use them in production scenarios.

Even for partners who choose to ignore this guidance and do so anyway, I think this is a relatively easy fixup for them.